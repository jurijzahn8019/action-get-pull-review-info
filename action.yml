name: "Get pull requets review info"
description: Action to retrieve review data for a pull request.
inputs:
  token:
    description: Github repo token, usually `secrets.GITHUB_TOKEN`
    required: true
  owner:
    description: "The organization/user owns the repository, default from event data"
    required: false
  number:
    description: "Pull request number to look for, default from event data"
    required: false
  repo:
    description: |
      the repo name (reponame in the owner scope without owner path)
      where the pull request is located, default from event data
    required: false
  collaborators:
    description: |
      This can be one or multiple of: 'COLLABORATOR', 'CONTRIBUTOR', 'FIRST_TIME_CONTRIBUTOR', 'FIRST_TIMER', 'MEMBER', 'OWNER', 'NONE'

      Multiple values are to be separated by `,` (comma). For example: `COLLABORATOR,CONTRIBUTOR`, Defaults to All possible

      This can be used to perform extra calculations only for given collaborator associations
    required: false
outputs:
  requested:
    description: |
      Count if the Reamining Review Requestes of the Pull Reqeuest.
      This will be 0 when all requested Peoples (Teams) have given their
      Reviews
  reviews:
    description: |
      Count of the given reviews in all states.
  approved:
    description: |
      Overal Count of the Reviews with Approvals
  changes_requested:
    description: |
      Overal Count of the Reviews wich Requests changes
  commented:
    description: |
      Overal Count of the Reviews with only comments
  dismissed:
    description: |
      Overal Count of the Dismissed Reviews
  pending:
    description: |
      Overal Count of the Pending Reviews

  approved_collaborators:
    description: Detail status by Allowed Author Associations
  approved_collaborator:
    description: Detail status by Author Association
  approved_contributor:
    description: Detail status by Author Association
  approved_first_time_contributor:
    description: Detail status by Author Association
  approved_first_timer:
    description: Detail status by Author Association
  approved_member:
    description: Detail status by Author Association
  approved_owner:
    description: Detail status by Author Association
  approved_none:
    description: Detail status by Author Association

  changes_requested_collaborators:
    description: Detail status by Allowed Author Associations
  changes_requested_collaborator:
    description: Detail status by Author Association
  changes_requested_contributor:
    description: Detail status by Author Association
  changes_requested_first_time_contributor:
    description: Detail status by Author Association
  changes_requested_first_timer:
    description: Detail status by Author Association
  changes_requested_member:
    description: Detail status by Author Association
  changes_requested_owner:
    description: Detail status by Author Association
  changes_requested_none:
    description: Detail status by Author Association

  commented_collaborators:
    description: Detail status by Allowed Author Associations
  commented_collaborator:
    description: Detail status by Author Association
  commented_contributor:
    description: Detail status by Author Association
  commented_first_time_contributor:
    description: Detail status by Author Association
  commented_first_timer:
    description: Detail status by Author Association
  commented_member:
    description: Detail status by Author Association
  commented_owner:
    description: Detail status by Author Association
  commented_none:
    description: Detail status by Author Association

  dismissed_collaborators:
    description: Detail status by Allowed Author Associations
  dismissed_collaborator:
    description: Detail status by Author Association
  dismissed_contributor:
    description: Detail status by Author Association
  dismissed_first_time_contributor:
    description: Detail status by Author Association
  dismissed_first_timer:
    description: Detail status by Author Association
  dismissed_member:
    description: Detail status by Author Association
  dismissed_owner:
    description: Detail status by Author Association
  dismissed_none:
    description: Detail status by Author Association

  pending_collaborators:
    description: Detail status by Allowed Author Associations
  pending_collaborator:
    description: Detail status by Author Association
  pending_contributor:
    description: Detail status by Author Association
  pending_first_time_contributor:
    description: Detail status by Author Association
  pending_first_timer:
    description: Detail status by Author Association
  pending_member:
    description: Detail status by Author Association
  pending_owner:
    description: Detail status by Author Association
  pending_none:
    description: Detail status by Author Association

runs:
  using: "node12"
  main: "dist/index.cjs.js"
