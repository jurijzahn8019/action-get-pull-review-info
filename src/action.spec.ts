/* eslint-disable @typescript-eslint/no-explicit-any */
import { getInput, info, setFailed, setOutput } from "@actions/core";
import { getOctokit } from "@actions/github";
import { run } from "./action";

jest.mock("@actions/core");
jest.mock("@actions/github");

const getInputMock = getInput as jest.MockedFunction<typeof getInput>;
const setFailedMock = setFailed as jest.MockedFunction<typeof setFailed>;
const getOctokitMock = getOctokit as jest.MockedFunction<typeof getOctokit>;
const infoMock = info as jest.MockedFunction<typeof info>;
const setOutputMock = setOutput as jest.MockedFunction<typeof setOutput>;

const octokit = {
  graphql: jest.fn(),
};

describe("action", () => {
  beforeEach(() => {
    process.env.GITHUB_EVENT_PATH = "fixtures/review.submitted.event.json";

    getOctokitMock.mockReturnValue(octokit as any);
    getInputMock.mockReturnValueOnce("THE TOKEN");
    octokit.graphql.mockResolvedValue({
      repository: {
        pullRequest: {
          mergeable: "MERGEABLE",
          mergeStateStatus: "CLEAN",
          reviewRequests: {
            nodes: [
              {
                requestedReviewer: {
                  name: "bim44666",
                },
              },
            ],
          },
          reviews: {
            nodes: [
              {
                author: {
                  name: "bim44666",
                },
                updatedAt: "2020-12-16T14:15:43Z",
                authorAssociation: "NONE",
                state: "CHANGES_REQUESTED",
              },
              {
                author: {
                  name: "foobar5344",
                },
                updatedAt: "2020-12-16T14:17:14Z",
                authorAssociation: "FIRST_TIMER",
                state: "CHANGES_REQUESTED",
              },
              {
                author: {
                  name: "bazbar634",
                },
                updatedAt: "2020-12-16T14:55:57Z",
                authorAssociation: "CONTRIBUTOR",
                state: "COMMENTED",
              },
              {
                author: {
                  name: "bazbar634",
                },
                updatedAt: "2020-12-17T07:10:16Z",
                authorAssociation: "CONTRIBUTOR",
                state: "DISMISSED",
              },
              {
                author: {
                  name: "bazbar634",
                },
                updatedAt: "2020-12-17T07:11:08Z",
                authorAssociation: "CONTRIBUTOR",
                state: "COMMENTED",
              },
              {
                author: {
                  name: "bazbar634",
                },
                updatedAt: "2020-12-17T07:51:47Z",
                authorAssociation: "CONTRIBUTOR",
                state: "CHANGES_REQUESTED",
              },
            ],
          },
        },
      },
    });
  });

  describe("run", () => {
    it("Happy Path", async () => {
      await expect(run()).resolves.toBeUndefined();

      expect(getInputMock).toHaveBeenCalledTimes(5);
      expect(getInputMock).toHaveBeenNthCalledWith(1, "token", {
        required: true,
      });
      expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN", {
        previews: ["merge-info"],
      });
      expect(octokit.graphql).toHaveBeenCalledWith({
        number: 743,
        owner: "otto-ec",
        repo: "assets_protecty",
        query: `
      query ($owner: String!, $repo: String!, $number: Int!) {
        repository(owner: $owner, name: $repo) {
          pullRequest(number: $number) {
            reviewRequests(last: 100) {
              nodes {
                requestedReviewer {
                  ... on User { name: login }
                  ... on Team { name }
                }
              }
            }
            reviews(last: 100) {
              nodes {
                author { name: login }
                updatedAt
                authorAssociation
                state
              }
            }
            mergeable
            mergeStateStatus
          }
        }
      }
    `,
      });

      expect(infoMock.mock.calls.map((c) => c.join(", "))).toMatchSnapshot(
        "Info",
      );
      expect(setOutputMock.mock.calls.map((c) => c.join(", "))).toMatchSnapshot(
        "setOutput",
      );
      expect(setFailedMock).not.toHaveBeenCalled();
    });

    it("Should fail due to missing input", async () => {
      process.env.GITHUB_EVENT_PATH = "foo/bar";

      await expect(run()).resolves.toBeUndefined();

      expect(setFailedMock).toHaveBeenCalledWith(
        "Failed to retrieve required parameters",
      );
    });

    it("Should use inputs", async () => {
      getInputMock.mockReturnValueOnce("the-owner");
      getInputMock.mockReturnValueOnce("the-repo");
      getInputMock.mockReturnValueOnce("666999");
      getInputMock.mockReturnValueOnce("FIRST_TIMER");

      await expect(run()).resolves.toBeUndefined();

      expect(octokit.graphql).toHaveBeenCalledWith({
        number: 666999,
        owner: "the-owner",
        repo: "the-repo",
        query: `
      query ($owner: String!, $repo: String!, $number: Int!) {
        repository(owner: $owner, name: $repo) {
          pullRequest(number: $number) {
            reviewRequests(last: 100) {
              nodes {
                requestedReviewer {
                  ... on User { name: login }
                  ... on Team { name }
                }
              }
            }
            reviews(last: 100) {
              nodes {
                author { name: login }
                updatedAt
                authorAssociation
                state
              }
            }
            mergeable
            mergeStateStatus
          }
        }
      }
    `,
      });

      expect(setFailedMock).not.toHaveBeenCalled();
      expect(setOutputMock.mock.calls.map((c) => c.join(", "))).toMatchSnapshot(
        "setOutput",
      );
    });

    it("Should handle no data", async () => {
      octokit.graphql.mockResolvedValue({
        repository: {
          pullRequest: {},
        },
      });

      await expect(run()).resolves.toBeUndefined();

      expect(setFailedMock).not.toHaveBeenCalled();
      expect(setOutputMock.mock.calls.map((c) => c.join(", "))).toMatchSnapshot(
        "setOutput",
      );
    });

    it("Should handle inaccessible data", async () => {
      octokit.graphql.mockResolvedValue({
        repository: {
          pullRequest: {
            mergeable: "UNKNOWN",
            mergeStateStatus: "DIRTY",
            reviewRequests: {
              nodes: [
                { requestedReviewer: null },
                { requestedReviewer: { name: "bim44666" } },
              ],
            },
            reviews: {
              nodes: [
                {
                  author: {
                    name: "bim44666",
                  },
                  updatedAt: "2020-12-16T14:15:43Z",
                  authorAssociation: "NONE",
                  state: "CHANGES_REQUESTED",
                },
                {
                  author: {
                    name: "foobar5344",
                  },
                  updatedAt: "2020-12-16T14:17:14Z",
                  authorAssociation: "FIRST_TIMER",
                  state: "CHANGES_REQUESTED",
                },
                {
                  author: {
                    name: "bazbar634",
                  },
                  updatedAt: "2020-12-16T14:55:57Z",
                  authorAssociation: "CONTRIBUTOR",
                  state: "COMMENTED",
                },
                {
                  author: {
                    name: "bazbar634",
                  },
                  updatedAt: "2020-12-17T07:10:16Z",
                  authorAssociation: "CONTRIBUTOR",
                  state: "DISMISSED",
                },
                {
                  author: {
                    name: "bazbar634",
                  },
                  updatedAt: "2020-12-17T07:11:08Z",
                  authorAssociation: "CONTRIBUTOR",
                  state: "COMMENTED",
                },
                {
                  author: {
                    name: "bazbar634",
                  },
                  updatedAt: "2020-12-17T07:51:47Z",
                  authorAssociation: "CONTRIBUTOR",
                  state: "CHANGES_REQUESTED",
                },
              ],
            },
          },
        },
      });

      await expect(run()).resolves.toBeUndefined();

      expect(setFailedMock).not.toHaveBeenCalled();
      expect(setOutputMock.mock.calls.map((c) => c.join(", "))).toMatchSnapshot(
        "setOutput",
      );
    });
  });
});
