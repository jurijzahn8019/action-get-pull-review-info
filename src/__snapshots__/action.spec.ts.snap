// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action run Happy Path: Info 1`] = `
Array [
  "Pull request info",
  "Reviews Given: 2",
  "Requested Remaining: 1",
  "approved: 0",
  "approved for Allowed Collaborators: 0",
  "changes_requested: 2",
  "changes_requested for Allowed Collaborators: 2",
  "commented: 0",
  "commented for Allowed Collaborators: 0",
  "dismissed: 0",
  "dismissed for Allowed Collaborators: 0",
  "pending: 0",
  "pending for Allowed Collaborators: 0",
  "Done",
]
`;

exports[`action run Happy Path: setOutput 1`] = `
Array [
  "reviews, 2",
  "requested, 1",
  "approved, 0",
  "approved_collaborator, 0",
  "approved_contributor, 0",
  "approved_first_time_contributor, 0",
  "approved_first_timer, 0",
  "approved_member, 0",
  "approved_owner, 0",
  "approved_none, 0",
  "approved_collaborators, 0",
  "changes_requested, 2",
  "changes_requested_collaborator, 0",
  "changes_requested_contributor, 1",
  "changes_requested_first_time_contributor, 0",
  "changes_requested_first_timer, 1",
  "changes_requested_member, 0",
  "changes_requested_owner, 0",
  "changes_requested_none, 0",
  "changes_requested_collaborators, 2",
  "commented, 0",
  "commented_collaborator, 0",
  "commented_contributor, 0",
  "commented_first_time_contributor, 0",
  "commented_first_timer, 0",
  "commented_member, 0",
  "commented_owner, 0",
  "commented_none, 0",
  "commented_collaborators, 0",
  "dismissed, 0",
  "dismissed_collaborator, 0",
  "dismissed_contributor, 0",
  "dismissed_first_time_contributor, 0",
  "dismissed_first_timer, 0",
  "dismissed_member, 0",
  "dismissed_owner, 0",
  "dismissed_none, 0",
  "dismissed_collaborators, 0",
  "pending, 0",
  "pending_collaborator, 0",
  "pending_contributor, 0",
  "pending_first_time_contributor, 0",
  "pending_first_timer, 0",
  "pending_member, 0",
  "pending_owner, 0",
  "pending_none, 0",
  "pending_collaborators, 0",
]
`;

exports[`action run Should handle no data: setOutput 1`] = `
Array [
  "reviews, 0",
  "requested, 0",
  "approved, 0",
  "approved_collaborator, 0",
  "approved_contributor, 0",
  "approved_first_time_contributor, 0",
  "approved_first_timer, 0",
  "approved_member, 0",
  "approved_owner, 0",
  "approved_none, 0",
  "approved_collaborators, 0",
  "changes_requested, 0",
  "changes_requested_collaborator, 0",
  "changes_requested_contributor, 0",
  "changes_requested_first_time_contributor, 0",
  "changes_requested_first_timer, 0",
  "changes_requested_member, 0",
  "changes_requested_owner, 0",
  "changes_requested_none, 0",
  "changes_requested_collaborators, 0",
  "commented, 0",
  "commented_collaborator, 0",
  "commented_contributor, 0",
  "commented_first_time_contributor, 0",
  "commented_first_timer, 0",
  "commented_member, 0",
  "commented_owner, 0",
  "commented_none, 0",
  "commented_collaborators, 0",
  "dismissed, 0",
  "dismissed_collaborator, 0",
  "dismissed_contributor, 0",
  "dismissed_first_time_contributor, 0",
  "dismissed_first_timer, 0",
  "dismissed_member, 0",
  "dismissed_owner, 0",
  "dismissed_none, 0",
  "dismissed_collaborators, 0",
  "pending, 0",
  "pending_collaborator, 0",
  "pending_contributor, 0",
  "pending_first_time_contributor, 0",
  "pending_first_timer, 0",
  "pending_member, 0",
  "pending_owner, 0",
  "pending_none, 0",
  "pending_collaborators, 0",
]
`;

exports[`action run Should use inputs: setOutput 1`] = `
Array [
  "reviews, 2",
  "requested, 1",
  "approved, 0",
  "approved_collaborator, 0",
  "approved_contributor, 0",
  "approved_first_time_contributor, 0",
  "approved_first_timer, 0",
  "approved_member, 0",
  "approved_owner, 0",
  "approved_none, 0",
  "approved_collaborators, 0",
  "changes_requested, 2",
  "changes_requested_collaborator, 0",
  "changes_requested_contributor, 1",
  "changes_requested_first_time_contributor, 0",
  "changes_requested_first_timer, 1",
  "changes_requested_member, 0",
  "changes_requested_owner, 0",
  "changes_requested_none, 0",
  "changes_requested_collaborators, 1",
  "commented, 0",
  "commented_collaborator, 0",
  "commented_contributor, 0",
  "commented_first_time_contributor, 0",
  "commented_first_timer, 0",
  "commented_member, 0",
  "commented_owner, 0",
  "commented_none, 0",
  "commented_collaborators, 0",
  "dismissed, 0",
  "dismissed_collaborator, 0",
  "dismissed_contributor, 0",
  "dismissed_first_time_contributor, 0",
  "dismissed_first_timer, 0",
  "dismissed_member, 0",
  "dismissed_owner, 0",
  "dismissed_none, 0",
  "dismissed_collaborators, 0",
  "pending, 0",
  "pending_collaborator, 0",
  "pending_contributor, 0",
  "pending_first_time_contributor, 0",
  "pending_first_timer, 0",
  "pending_member, 0",
  "pending_owner, 0",
  "pending_none, 0",
  "pending_collaborators, 0",
]
`;
